File: programs/roundfi-core/src/instructions/finalize_draw.rs
//! `finalize_draw` — mint the payout-order permutation for a sorteio pool
//! (ADR pool_v2). Permissionless, exactly once per pool.
//!
//! When a pool with `ordering_policy == ORDERING_SORTEIO` fills (last
//! `join_pool` flips it Active), payouts are UNREACHABLE — the three
//! payout instructions fail `DrawRequired` — until someone runs this. The
//! app fires it right after the activating join; because it's
//! permissionless (same trust model as `crank_payout`: no funds move
//! anywhere but where the protocol already says they go), any member can
//! unstick a pool whose creator forgot.
//!
//! **Seed (v1-canary, honest limitation).** sha256(pool ‖ clock.slot ‖
//! clock.unix_timestamp ‖ members_target). The finalize caller can grind
//! the timing of this transaction (~1 slot granularity) to nudge the
//! permutation — bounded, but real. Acceptable for the devnet canary; a
//! VRF-bound seed (Switchboard/ORAO) replaces this before mainnet, which
//! is why the seed is STORED on the DrawResult: the upgrade only changes
//! where the bytes come from, and audits can always re-derive
//! `order = draw_slot_order(seed, n)`.
//!
//! **Window re-anchor.** The cycle-0 clock started at activation, but a
//! sorteio pool is only operable once drawn. If the draw lands late, the
//! contemplated member's claim window would be silently eaten — the exact
//! unfairness class SEV-053 closed. Same cure: re-anchor
//! `next_cycle_at = max(next_cycle_at, now + cycle_duration)`, so an
//! immediate finalize is a no-op and a late one opens a full window.

use anchor_lang::prelude::*;
use solana_program::hash::hashv;

use crate::constants::{ORDERING_SORTEIO, SEED_DRAW_RESULT};
use crate::error::RoundfiError;
use crate::math::draw_slot_order_checked;
use crate::state::{DrawResult, Pool, PoolStatus};

#[derive(Accounts)]
pub struct FinalizeDraw<'info> {
    /// Permissionless caller — pays the DrawResult rent.
    #[account(mut)]
    pub caller: Signer<'info>,

    #[account(
        mut,
        constraint = pool.ordering_policy == ORDERING_SORTEIO @ RoundfiError::DrawNotReady,
        constraint = pool.status == PoolStatus::Active as u8 @ RoundfiError::DrawNotReady,
        constraint = pool.members_joined == pool.members_target @ RoundfiError::DrawNotReady,
    )]
    pub pool: Box<Account<'info, Pool>>,

    /// `init` (not init_if_needed) makes the draw single-shot: a second
    /// finalize collides on the existing PDA and reverts — nobody can
    /// re-roll an unfavorable permutation.
    #[account(
        init,
        payer = caller,
        space = DrawResult::SIZE,
        seeds = [SEED_DRAW_RESULT, pool.key().as_ref()],
        bump,
    )]
    pub draw: Box<Account<'info, DrawResult>>,

    pub system_program: Program<'info, System>,
}

pub fn handler(ctx: Context<FinalizeDraw>) -> Result<()> {
    let pool = &mut ctx.accounts.pool;
    let draw = &mut ctx.accounts.draw;
    let clock = Clock::get()?;
    let n = pool.members_target;

    // v1-canary entropy — see module docs for the grinding caveat + the
    // VRF upgrade path. Stored so the permutation stays re-derivable.
    let seed = hashv(&[
        pool.key().as_ref(),
        &clock.slot.to_le_bytes(),
        &clock.unix_timestamp.to_le_bytes(),
        &[n],
    ])
    .to_bytes();

    let mut order = [0u8; crate::constants::MAX_MEMBERS as usize];
    // Bijection over 0..n guaranteed by roundfi_math (unit + fuzz tested);
    // the wrapper maps MathError → RoundfiError at the boundary.
    draw_slot_order_checked(&seed, &mut order[..n as usize])?;

    draw.pool = pool.key();
    draw.seed = seed;
    draw.order = order;
    draw.members_target = n;
    draw.bump = ctx.bumps.draw;

    // SEV-053-pattern re-anchor: the pool becomes operable NOW, so cycle
    // 0 gets a full window from now (immediate finalize ⇒ max picks the
    // activation schedule, no drift).
    pool.next_cycle_at = pool.next_cycle_at.max(
        clock
            .unix_timestamp
            .checked_add(pool.cycle_duration)
            .ok_or(error!(RoundfiError::MathOverflow))?,
    );

    msg!(
        "roundfi-core: draw finalized pool={} n={} seed={:?} (payout order minted; cycle-0 window re-anchored)",
        pool.key(),
        n,
        &seed[..8],
    );
    Ok(())
}

File: app/src/lib/finalize-draw.ts
/**
 * IDL-free encoder + sender for `roundfi-core::finalize_draw` (ADR
 * pool_v2 — sorteio ordering policy).
 *
 * Permissionless, single-shot: when a sorteio pool fills, anyone can run
 * this to mint the pool's DrawResult PDA (the payout-order permutation).
 * Payouts on a sorteio pool are unreachable (`DrawRequired`) until it
 * runs; a second call collides on the PDA `init` — nobody re-rolls an
 * unfavorable order. Same trust model as `crank_payout`: no funds move.
 *
 * Account order MUST match `FinalizeDraw<'info>` in
 * programs/roundfi-core/src/instructions/finalize_draw.rs (4 accounts).
 */

import {
  Connection,
  PublicKey,
  SystemProgram,
  Transaction,
  TransactionInstruction,
} from "@solana/web3.js";

import { drawResultPda } from "@roundfi/sdk/pda";

import { DEVNET_PROGRAM_IDS } from "./devnet";
import { simulateOrThrow } from "./simulateTx";

// sha256("global:finalize_draw")[:8] — precomputed.
//   $ node -e 'console.log(require("crypto").createHash("sha256")
//                .update("global:finalize_draw").digest()
//                .subarray(0,8).toString("hex"))'
//   → 7009ea5e63b00cb5
const FINALIZE_DRAW_DISCRIMINATOR = Buffer.from([0x70, 0x09, 0xea, 0x5e, 0x63, 0xb0, 0x0c, 0xb5]);

export interface BuildFinalizeDrawIxArgs {
  /** Pool PDA — must be a full, Active sorteio pool. */
  pool: PublicKey;
  /** Permissionless caller — signs + pays the DrawResult rent. */
  caller: PublicKey;
}

/** Build the raw `finalize_draw()` instruction (no args, 4 accounts). */
export function buildFinalizeDrawIx(args: BuildFinalizeDrawIxArgs): TransactionInstruction {
  const core = DEVNET_PROGRAM_IDS.core;
  const [draw] = drawResultPda(core, args.pool);

  return new TransactionInstruction({
    programId: core,
    data: FINALIZE_DRAW_DISCRIMINATOR,
    keys: [
      { pubkey: args.caller, isSigner: true, isWritable: true },
      { pubkey: args.pool, isSigner: false, isWritable: true },
      { pubkey: draw, isSigner: false, isWritable: true },
      { pubkey: SystemProgram.programId, isSigner: false, isWritable: false },
    ],
  });
}

export interface SendFinalizeDrawArgs extends BuildFinalizeDrawIxArgs {
  connection: Connection;
  sendTransaction: (tx: Transaction, connection: Connection) => Promise<string>;
}

/** Simulate-then-sign `finalize_draw`. Returns the tx signature. */
export async function sendFinalizeDraw(args: SendFinalizeDrawArgs): Promise<string> {
  const ix = buildFinalizeDrawIx(args);
  const tx = new Transaction().add(ix);
  tx.feePayer = args.caller;
  const { blockhash } = await args.connection.getLatestBlockhash("confirmed");
  tx.recentBlockhash = blockhash;
  await simulateOrThrow(args.connection, tx);
  return args.sendTransaction(tx, args.connection);
}
